### PR TITLE
Zoom, Rendering Notification, and Misc Cleanup

### PR DESCRIPTION
**Pull request by DXCanas**
Solves the console errors described in https://github.com/learningequality/kolibri/pull/1888#pullrequestreview-51858828

### CODE REVIEW

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 1]
```diff
@@ -13,8 +13,12 @@
       <mat-svg v-if="isFullscreen" class="icon" category="navigation" name="fullscreen_exit"/>
       <mat-svg v-else class="icon" category="navigation" name="fullscreen"/>
     </icon-button>
+
+    <icon-button text="Zoom In" @click="zoomIn()"/>
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Unwrapped, unstyled. WIll be soon!

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 2]
```diff
@@ -13,8 +13,12 @@
...
+    <icon-button text="Zoom out" @click="zoomOut()"/>
+
     <div ref="pdfcontainer" class="pdfcontainer" @scroll="checkPages">
-      <progress-bar v-if="loading" class="progress-bar" :show-percentage="true" :progress="progress"/>
+      <progress-bar v-if="documentLoading" class="progress-bar" :show-percentage="true" :progress="progress"/>
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Just using a computed prop here. Thinking about going back to manually setting to have the page rendered on arrival

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 3]
```diff
@@ -63,48 +75,83 @@
...
+      },
     },
     methods: {
       toggleFullscreen() {
-        if (this.isFullscreen) {
-          if (this.fullscreenAllowed) {
-            ScreenFull.toggle(this.$refs.container);
-          }
-          this.isFullscreen = false;
+        if (this.fullscreenAllowed) {
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue:108]: A minor fullscreen bugfix. Might be best to break this into a separate PR :man_shrugging:

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 4]
```diff
@@ -63,48 +75,83 @@
...
+      zoomIn() {
+        this.scale += 0.1;
       },
-      getPage(pageNum) {
-        return this.pdfDocument.getPage(pageNum);
+      zoomOut() {
+        this.scale -= 0.1;
       },
-      displayPage(pdfPage) {
+      getPage(pageNum, firstRender = false) {
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Probably ditching the firstRenderOption? Or just renaming `setupIntialPageScale` to `setupZoom`

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 5]
```diff
@@ -63,48 +75,83 @@
...
+
+        const isMobile = this.windowSize.breakpoint === 0;
+
+        if (isMobile) {
+          this.scale = this.elSize.width / pdfPageWidth;
+        } else {
+          this.scale = this.elSize.height / pdfPageHeight;
+        }
+      },
+      startRender(pdfPage) {
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Renamed it `startRender` to avoid confusion between `display` and `show`

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 6]
```diff
@@ -175,49 +220,69 @@
       pageRef(index) {
         return `pdfPage-${index}`;
       },
-      checkPages() {
+      // debouncing so we're not de/re-render many pages unnecessarily
+      checkPages: debounce(function() {
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Added the debounce here

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 7]
```diff
@@ -175,49 +220,69 @@
...
+        // Loop through all pages, show ones that are in the display window, hide ones that aren't
+        for (let i = 1; i <= this.totalPages; i++) {
           if (i < topPageNum - pageDisplayWindow || i > bottomPageNum + pageDisplayWindow) {
             this.hidePage(i);
           } else {
             this.showPage(i);
           }
         }
-      },
+      }, renderDebounceTime),
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue:251]: Set outside of the `exports` scope

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 8]
```diff
@@ -175,49 +220,69 @@
...
     },
     watch: {
       scrollPos: 'checkPages',
+      scale(newScale, oldScale) {
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue:255]: The `scale` watcher

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 9]
```diff
@@ -175,49 +220,69 @@
...
-        }
-      ).then(pdfDocument => {
-        this.loading = false;
-        this.totalPages = pdfDocument.numPages;
-        // Track the pdf document
-        this.pdfDocument = pdfDocument;
-        // Begin retrieving the first page
-        return this.getPage(1);
-      });
+      this.loadPdfPromise = PDFJSLib.getDocument(this.defaultFile.storage_url);
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Stripped created down to be a little simpler. Moved a lot of the `.then` logic down into mounted. Haven't noticed any issues, but please advise

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 10]
```diff
@@ -175,49 +220,69 @@
...
     },
     mounted() {
-      this.pdfloadingPromise.then(() => {
-        this.showPage(1);
-        this.$emit('startTracking');
+      if (this.fullscreenAllowed) {
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: A part of the fullscreen bugfix. It wasn't updating the text/icon if you used the escape button

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 11]
```diff
@@ -175,49 +220,69 @@
...
+        ScreenFull.onchange(() => {
+          this.isFullscreen = ScreenFull.isFullscreen;
+        });
+      }
+
+      // pass callback to update loading bar
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: This is the "supported" way to specify a progress callback

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 12]
```diff
@@ -175,49 +220,69 @@
...
+
+      // Retrieve the document and its corresponding object
+      this.loadPdfPromise.then(pdfDocument => {
+        this.pdfDocument = pdfDocument;
+        this.totalPages = pdfDocument.numPages;
+        this.pdfPages = {};
+
+        // Retrieve the first Page object
+        this.getPage(1).then(pdfPage => {
+          this.setupInitialPageScale(pdfPage);
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Where the page scale is specified

File: package.json [hunk 13]
```diff
@@ -46,7 +46,7 @@
     "material-design-icons": "^3.0.1",
     "purecss": "^0.6.2",
     "rest": "1.3.2",
-    "screenfull": "3.0.2",
+    "screenfull": "3.3.1",
```
Comment (DXCanas) [inline package.json:49]: Tested a bit, not noticing any issues

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 14]
```diff
@@ -63,48 +75,83 @@
...
-            if (this.pageHeight === 0 && this.pageWidth === 0) {
-              this.pageHeight = viewport.height;
-              this.pageWidth = viewport.width;
-            }
-            canvas.width = viewport.width;
-            canvas.height = viewport.height;
-            const ctx = canvas.getContext('2d');
+
+            // define canvas and dummy blank page dimensions
+            canvas.width = this.pageWidth = viewport.width;
```
Comment (DXCanas) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue:143]: both canvas and pageWidth are being defined all the time now.

**Review by DXCanas — COMMENTED** [review]
Quick comments on some changes I've made!

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue (hunk 11 above)
Comment (rtibbles) [reply kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: I think you should add this within the `created` handler, rather than `mounted` - data is available to be modified, it just won't be used for rendering yet.

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue [hunk 15]
```diff
@@ -175,49 +219,69 @@
...
+      };
+
+      // Retrieve the document and its corresponding object
+      this.loadPdfPromise.then(pdfDocument => {
+        this.pdfDocument = pdfDocument;
+        this.totalPages = pdfDocument.numPages;
+        this.pdfPages = {};
+
+        // Retrieve the first Page object
+        this.getPage(1).then(pdfPage => {
```
Comment (rtibbles) [inline kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: Note that you are now deferring getting the first page until the component has mounted - this should be moved back into the `created` handler, so that we are not waiting on DOM mounting for this operation. The only thing we need to wait for DOM mounting for is actually rendering into the page.

**Comment by rtibbles** [comment]
On tablet landscape mode, I think I would expect the PDF to expand horizontally more by default?
image

Could be a bit wider? And getting overflow of the Author field below in tablet portrait mode:
image


Look out for button overlap on mobile portrait:
image

**Review by rtibbles — CHANGES_REQUESTED** [review]
Renders smoothly, and with far fewer warnings about request animation frame delays, so I think the debouncing is working well.

Couple of comments here, plus some UI comments with screenshots.

File: kolibri/plugins/document_pdf_render/assets/src/views/index.vue (hunk 15 above)
Comment (DXCanas) [reply kolibri/plugins/document_pdf_render/assets/src/views/index.vue]: I was experiencing a flicker with the loading bar, and thought it might be because it'd been loading prior to being mounted. Thought moving it to mounted would fix that, but I'm not seeing any huge benefit

**Comment by DXCanas** [comment]
> On tablet landscape mode, I think I would expect the PDF to expand horizontally more by default?

So it'd start of rendered at full width on both tablet and phone? (landscape and portrait)

Do we just want to avoid full page height all together? I took this approach due to feedback from khang

> Could be a bit wider?

I agree, though @jessica mentioned that it got annoying for users to scroll the page when the pdf was taking up so much width. They had nowhere to grab the page itself.

> Look out for button overlap on mobile portrait:

Forgot to mention that they're going to be fixed to the bottom right. Jessica thought just icons would be best. Do you have opinions on that?

**Comment by rtibbles** [comment]
It's just a lot of grey space for a very small area of content. I can see the advantage of allowing people to see a page at a time, and we would want to avoid using the full width on ultra wide screens, but I think on 4:3 aspect ratio we should probably be using the full width.

On mobile users should be able to press on the page itself to scroll, and as long as we are not going so wide as to push out the scroll bar, should be fine on desktop?

No strong opinions on the buttons, bottom right display works for me, consistent with Chrome built in PDF display.

@khangmach @jtamiace input appreciated.

**Comment by khangmach** [comment]
Can we have the background of the pdf container be something darker than this current #f9f9f9? Maybe #212121? Having a background gives something to compare the width of the PDF to and makes margins look more visible. Buttons bottom right sound fine to me.

**Comment by jtamiace** [comment]
**Mobile**
I think there needs to be enough space on either side of the PDF renderer in order to allow the user to scroll vertically with their finger, so I'm okay with the grey space on the sides. Reading a PDF on mobile is going to be a painful experience no matter what, and there is the fullscreen mode that they can enable. 

Agree that the PDF itself should be rendered full-width also.
Also, using icon buttons for zoom in/out goes against the style guide but I think this is a very reasonable exception. Bottom right sounds fine here because you can see the entire renderer.

**Tablet**
I think it's also reasonable for the PDF to render full-width, both landscape and portrait.

Icon buttons might be good to align top/right at this breakpoint since the PDF renderer may extend offscreen at the bottom and hide the buttons.

**Desktop**
I don't think it's a good idea to render full-width OR fit to the height of the screen. The document is unreadable when it's that small, and also when it's too big. Would suggest that the document by default loads to some size in between. I don't think the user loses much by not seeing 100% of every page.

Icon buttons might be good to align top/right at this breakpoint since the PDF renderer may extend offscreen at the bottom and hide the buttons.

**Comment by jtamiace** [comment]
Regarding background color, @DXCanas did you remove the dark grey that you showed me yesterday? It looked nice

**Comment by rtibbles** [comment]
For desktop, we can just default to the document scale of `1.0`?

@jtamiace get your point about scrolling now, it's about scrolling the Kolibri page, not the PDF page.

**Comment by rtibbles** [comment]
I think no matter where they zoom buttons are placed, they should be fixed inside the PDF container and kept visible. I think the intention is for the height of the container itself not to increase beyond the height of the screen so they should be kept visible?

**Comment by jtamiace** [comment]
>  I think the intention is for the height of the container itself not to increase beyond the height of the screen so they should be kept visible?

If it is possible for the container to be scaled to browser height, bottom-right would work for consistency. It just looks awkward on desktop currently, where the container takes up about 50-60% of the available screen height.

screen shot 2017-08-01 at 10 16 23 am

**Comment by DXCanas** [comment]
@jtamiace @khangmach this is all still WIP. Was mostly for code review. I'll get all the suggestions in by the end

**Comment by DXCanas** [comment]
So, I've got:
- [x] Darker Background. Might wait on this until I merge latest develop so that I get the new colors(?)
- [x] Make it a bit taller but not much.
    - @jtamiace Calculating how much of the display we have while taking into account the nav, tabs, and breadcrumbs is difficult because they take up a different proportion of the screen depending on the height of the display. They're also not defined proportionally, they're defined using pixels, Right now, they're being guesstimated - I can guesstimate some more
- [x] PDF is full-width in every scenario BUT full size desktop, where it's just rendered to scale of the page
    - Full height is never a thing
- [X] Align zoom buttons to bottom on mobile
- [x] Align zoom button to top on desktop (unless you can find the magic height that's not awk and doesn't ever fall off the fold
- [x] Make sure pages to leave the container and cover the author
- [ ] Loading Message

Am I missing anything?

**Comment by DXCanas** [comment]
Update: Currently, there's a min height on the pdf viewer's box, so landscape tablet does go below the fold. 

Removing it makes it look ridiculous, though:
image

Gonna go with top-align in these cases

**Comment by jtamiace** [comment]
Might help the buttons pop out a bit more if you add drop shadow? Same one that appears on the raised buttons

**Comment by DXCanas** [comment]
@jtamiace I had issues adding a drop shadow to keen's UI buttons, but I'll give it a shot.

**Comment by DXCanas** [comment]
OK! I think it's ready for review

**Comment by DXCanas** [comment]
pdfdemo
@jtamiace @khangmach 

UI buttons don't lend themselves well to drop shadows, that's why they're not there @jtamiace

**Review by rtibbles — APPROVED** [review]
